src/kvstore.c

```diff
@@ -962,6 +1074,165 @@ kvstoreMetadata *kvstoreGetMetadata(kvstore *kvs) {
+    return es;
+}
+
+/* Empty an expiration store (clear all entries but keep the structure) */
+void estoreEmpty(estore *es) {
+    if (es == NULL) return;
+
+    for (int i = 0; i < es->num_buckets; i++) {
+        ebDestroy(&es->ebArray[i], es->bucket_type, NULL);
+        es->ebArray[i] = ebCreate();
+    }
+
+    es->count = 0;
+}
+
+int estoreIsEmpty(estore *es) {
+    return es->count == 0;
+}
+
+/* Get the first non-empty bucket index in the estore */
+int estoreGetFirstNonEmptyBucket(estore *es) {
+    if (es->num_buckets == 1 || estoreSize(es) == 0)
+        return 0;
+    return fwTreeFindFirstNonEmpty(es->buckets_sizes);
+}
+
+/* Get the next non-empty bucket index after the given index */
+int estoreGetNextNonEmptyBucket(estore *es, int eidx) {
+    if (es->num_buckets == 1) {
+        assert(eidx == 0);
+        return -1;
+    }
+    return fwTreeFindNextNonEmpty(es->buckets_sizes, eidx);
+}
+
+/* Release an expiration store (free all memory) */
+void estoreRelease(estore *es) {
+    if (es == NULL) return;
+
+    for (int i = 0; i < es->num_buckets; i++) {
+        if (es->ebArray[i])
+            ebDestroy(&es->ebArray[i], es->bucket_type, NULL);
+    }
+    fwTreeDestroy(es->buckets_sizes);
+    zfree(es->ebArray);
+    zfree(es);
+}
+
+void estoreActiveExpire(estore *es, int eidx, ExpireInfo *info) {
+    ebuckets *eb = estoreGetBuckets(es, eidx);
+    uint64_t before = ebGetTotalItems(*eb, es->bucket_type);
+    ebExpire(eb, es->bucket_type, info);
+    /* If items expired (or updated), update the BIT and estore count */
+    if (info->itemsExpired) {
+        uint64_t diff = before - ebGetTotalItems(*eb, es->bucket_type);
+        fwTreeUpdate(es->buckets_sizes, eidx, (long long) diff);
+        es->count -= diff;
+    }
+}
+
+/* Add item to estore with the given expiration time. The item must has
+ * expireMeta already allocated. */
+void estoreAdd(estore *es, int eidx, eItem item, uint64_t when) {
+    debugAssert(es != NULL && item != NULL);
+
+    /* currently only used by hash field expiration. Verify it has expireMeta */
+    debugAssert((((robj *)item)->encoding == OBJ_ENCODING_LISTPACK_EX) ||
+                ((((robj *)item)->encoding == OBJ_ENCODING_HT) &&
+                 ((dict *) ((robj *)item)->ptr)->type == &mstrHashDictTypeWithHFE));
+
+    ebuckets *bucket = estoreGetBuckets(es, eidx);
+    if (ebAdd(bucket, es->bucket_type, item, when) == 0) {
+        es->count++;
+        fwTreeUpdate(es->buckets_sizes, eidx, 1);
+    }
+}
+
+uint64_t estoreRemove(estore *es, int eidx, eItem item) {
+    uint64_t expireTime;
+    debugAssert(es != NULL && item != NULL);
+
+    /* currently only used by hash field expiration. Verify it is hash with HFE */
+    kvobj *kv = (kvobj *) item;
+    if ( (kv->type != OBJ_HASH) ||
+         (kv->encoding == OBJ_ENCODING_LISTPACK) ||
+         ((kv->encoding == OBJ_ENCODING_HT) && (((dict *)kv->ptr)->type != &mstrHashDictTypeWithHFE)))
+        return EB_EXPIRE_TIME_INVALID;
```

Comment:
perhaps this should be assert

Reply:
i reverted the assert. It might be called blindly on non HFEs and should return gracefully.